File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/main/java/io/github/lishangbu/avalon/dataset/controller/StatController.java
package io.github.lishangbu.avalon.dataset.controller;

import io.github.lishangbu.avalon.dataset.entity.Stat;
import io.github.lishangbu.avalon.dataset.service.StatService;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/// 能力(Stat)控制器
///
/// @author lishangbu
/// @since 2026/3/19
@RestController
@RequestMapping("/stat")
@RequiredArgsConstructor
public class StatController {
    private final StatService statService;

    /// 分页条件查询能力
    ///
    /// @param pageable 分页参数
    /// @param stat     查询条件，支持 name/internalName 模糊查询
    /// @return 能力分页结果
    @GetMapping("/page")
    public Page<Stat> getStatPage(Pageable pageable, Stat stat) {
        return statService.getPageByCondition(stat, pageable);
    }

    /// 新增能力
    ///
    /// @param stat 能力实体
    /// @return 保存后的能力
    @PostMapping
    public Stat save(@RequestBody Stat stat) {
        return statService.save(stat);
    }

    /// 更新能力
    ///
    /// @param stat 能力实体
    /// @return 更新后的能力
    @PutMapping
    public Stat update(@RequestBody Stat stat) {
        return statService.update(stat);
    }

    /// 根据 ID 删除能力
    ///
    /// @param id 能力主键
    @DeleteMapping("/{id:\\d+}")
    public void deleteById(@PathVariable Long id) {
        statService.removeById(id);
    }

    /// 条件查询能力列表
    ///
    /// @param stat 查询条件，支持 name/internalName 模糊查询
    /// @return 能力列表
    @GetMapping("/list")
    public List<Stat> listStats(Stat stat) {
        return statService.listByCondition(stat);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/main/kotlin/io/github/lishangbu/avalon/dataset/controller/StatController.kt
package io.github.lishangbu.avalon.dataset.controller

import io.github.lishangbu.avalon.dataset.entity.Stat
import io.github.lishangbu.avalon.dataset.entity.dto.SaveStatInput
import io.github.lishangbu.avalon.dataset.entity.dto.StatSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.StatView
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateStatInput
import io.github.lishangbu.avalon.dataset.repository.StatRepository
import jakarta.validation.Valid
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.springframework.web.bind.annotation.*

/** 能力值控制器 */
@RestController
@RequestMapping("/stat")
class StatController(
    /** 能力值仓储*/
    private val statRepository: StatRepository,
) {
    /** 保存能力值*/
    @PostMapping
    fun save(
        @Valid
        @RequestBody command: SaveStatInput,
    ): StatView = statRepository.save(command.toEntity(), SaveMode.INSERT_ONLY).let(::reloadView)

    /** 更新能力值*/
    @PutMapping
    fun update(
        @Valid
        @RequestBody command: UpdateStatInput,
    ): StatView {
        requireStatEditable(command.id.toLong(), "能力值已设为只读，禁止修改")
        return statRepository.save(command.toEntity(), SaveMode.UPSERT).let(::reloadView)
    }

    /** 按 ID 删除能力值*/
    @DeleteMapping("/{id:\\d+}")
    fun deleteById(
        @PathVariable id: Long,
    ) {
        requireStatEditable(id, "能力值已设为只读，禁止删除")
        statRepository.deleteById(id)
    }

    /** 查询能力值列表*/
    @GetMapping("/list")
    fun listStats(
        @ModelAttribute specification: StatSpecification,
    ): List<StatView> = statRepository.listViews(specification)

    private fun reloadView(stat: Stat): StatView = requireNotNull(statRepository.loadViewById(stat.id)) { "未找到 ID=${stat.id} 对应的能力值" }

    private fun requireStatEditable(
        id: Long,
        errorMessage: String,
    ) {
        val stat = requireNotNull(statRepository.findNullable(id)) { "未找到 ID=$id 对应的能力值" }
        if (stat.readonly) {
            throw IllegalStateException(errorMessage)
        }
    }
}
